File: native_locator/src/virtualenvwrapper.rs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

use crate::locator::{Locator, LocatorResult};
use crate::messaging::PythonEnvironment;
use crate::utils::list_python_environments;
use crate::virtualenv;
use crate::{known::Environment, utils::PythonEnv};
use std::path::PathBuf;

#[cfg(windows)]
fn get_default_virtualenvwrapper_path(environment: &dyn Environment) -> Option<PathBuf> {
    // In Windows, the default path for WORKON_HOME is %USERPROFILE%\Envs.
    // If 'Envs' is not available we should default to '.virtualenvs'. Since that
    // is also valid for windows.
    if let Some(home) = environment.get_user_home() {
        let home = PathBuf::from(home).join("Envs");
        if home.exists() {
            return Some(home);
        }
        let home = PathBuf::from(home).join("virtualenvs");
        if home.exists() {
            return Some(home);
        }
    }
    None
}

#[cfg(unix)]
fn get_default_virtualenvwrapper_path(environment: &dyn Environment) -> Option<PathBuf> {
    if let Some(home) = environment.get_user_home() {
        let home = PathBuf::from(home).join("virtualenvs");
        if home.exists() {
            return Some(home);
        }
    }
    None
}

fn get_work_on_home_path(environment: &dyn Environment) -> Option<PathBuf> {
    // The WORKON_HOME variable contains the path to the root directory of all virtualenvwrapper environments.
    // If the interpreter path belongs to one of them then it is a virtualenvwrapper type of environment.
    if let Some(work_on_home) = environment.get_env_var("WORKON_HOME".to_string()) {
        if let Ok(work_on_home) = std::fs::canonicalize(work_on_home) {
            if work_on_home.exists() {
                return Some(work_on_home);
            }
        }
    }
    get_default_virtualenvwrapper_path(environment)
}

pub fn is_virtualenvwrapper(env: &PythonEnv, environment: &dyn Environment) -> bool {
    if env.path.is_none() {
        return false;
    }
    // For environment to be a virtualenvwrapper based it has to follow these two rules:
    // 1. It should be in a sub-directory under the WORKON_HOME
    // 2. It should be a valid virtualenv environment
    if let Some(work_on_home_dir) = get_work_on_home_path(environment) {
        if env.executable.starts_with(&work_on_home_dir) && virtualenv::is_virtualenv(env) {
            return true;
        }
    }

    false
}

pub struct VirtualEnvWrapper<'a> {
    pub environment: &'a dyn Environment,
}

impl VirtualEnvWrapper<'_> {
    pub fn with<'a>(environment: &'a impl Environment) -> VirtualEnvWrapper {
        VirtualEnvWrapper { environment }
    }
}

impl Locator for VirtualEnvWrapper<'_> {
    fn resolve(&self, env: &PythonEnv) -> Option<PythonEnvironment> {
        if is_virtualenvwrapper(env, self.environment) {
            return Some(PythonEnvironment {
                name: Some(
                    env.path
                        .clone()
                        .expect("env.path cannot be empty for virtualenv rapper")
                        .file_name()
                        .unwrap()
                        .to_string_lossy()
                        .to_string(),
                ),
                python_executable_path: Some(env.executable.clone()),
                version: env.version.clone(),
                category: crate::messaging::PythonEnvironmentCategory::Venv,
                sys_prefix_path: env.path.clone(),
                env_path: env.path.clone(),
                env_manager: None,
                project_path: None,
                python_run_command: Some(vec![env.executable.to_str().unwrap().to_string()]),
            });
        }
        None
    }

    fn find(&self) -> Option<LocatorResult> {
        let work_on_home = get_work_on_home_path(self.environment)?;
        let envs = list_python_environments(&work_on_home)?;
        let mut environments: Vec<PythonEnvironment> = vec![];
        envs.iter().for_each(|env| {
            if let Some(env) = self.resolve(env) {
                environments.push(env);
            }
        });

        if environments.is_empty() {
            None
        } else {
            Some(LocatorResult::Environments(environments))
        }
    }
}

File: native_locator/ts-async/src/virtualenvwrapper.ts
/* eslint-disable class-methods-use-this */
/* eslint-disable consistent-return */
/* eslint-disable camelcase */
import * as path from 'path';
import * as fs from 'fs-extra';
import { homedir } from 'os';
import { PythonEnvironmentCategory, type PythonEnvironment } from './messaging';
import type { Locator, LocatorResult } from './locator';
import { listPythonEnvironments, type PythonEnv } from './utils';
import { isVirtualenv } from './virtualenv';

async function get_default_virtualenvwrapper_path(): Pormise<string | null> {
    if (process.platform === 'win32') {
        // In Windows, the default path for WORKON_HOME is %USERPROFILE%\Envs.
        // If 'Envs' is not available we should default to '.virtualenvs'. Since that
        // is also valid for windows.
        const home = homedir();
        if (home) {
            let homePath = path.join(home, 'Envs');
            if (await fs.pathExists(homePath)) {
                return homePath;
            }
            homePath = path.join(home, 'virtualenvs');
            if (await fs.pathExists(homePath)) {
                return homePath;
            }
        }
    } else {
        const home = homedir();
        if (home) {
            const homePath = path.join(home, 'virtualenvs');
            if (await fs.pathExists(homePath)) {
                return homePath;
            }
        }
    }
    return null;
}

async function get_work_on_home_path(): Promise<string | null> {
    // The WORKON_HOME variable contains the path to the root directory of all virtualenvwrapper environments.
    // If the interpreter path belongs to one of them then it is a virtualenvwrapper type of environment.
    const work_on_home = process.env.WORKON_HOME;
    if (work_on_home) {
        const workOnHomePath = path.resolve(work_on_home);
        if (await fs.pathExists(workOnHomePath)) {
            return workOnHomePath;
        }
    }
    return get_default_virtualenvwrapper_path();
}

async function is_virtualenvwrapper(env: PythonEnv): Promise< boolean> {
    if (!env.path) {
        return false;
    }
    // For environment to be a virtualenvwrapper based it has to follow these two rules:
    // 1. It should be in a sub-directory under the WORKON_HOME
    // 2. It should be a valid virtualenv environment
    const work_on_home_dir = await get_work_on_home_path();
    if (work_on_home_dir && env.executable.startsWith(work_on_home_dir) && (await isVirtualenv(env))) {
        return true;
    }
    return false;
}

export class VirtualEnvWrapper  {
    async resolve(env: PythonEnv): Promise< PythonEnvironment | undefined> {
        if (await is_virtualenvwrapper(env)) {
            return {
                name: env.path && path.basename(env.path),
                python_executable_path: env.executable,
                version: env.version,
                category: PythonEnvironmentCategory.Venv,
                sys_prefix_path: env.path,
                env_path: env.path,
                python_run_command: [env.executable],
            };
        }
    }

    async find(): Promise< LocatorResult | undefined> {
        const work_on_home = await get_work_on_home_path();
        if (work_on_home) {
            const envs = await listPythonEnvironments(work_on_home) || [];
            const environments: PythonEnvironment[] = [];
            await Promise.all(envs.map(async (env) => {
                const resolvedEnv = await this.resolve(env);
                if (resolvedEnv) {
                    environments.push(resolvedEnv);
                }
            }));

            if (environments.length === 0) {
                return;
            }
            return { environments };
        }
    }
}
